### ISSUE TITLE
What is the use of register_preprocessor in engine.rb?

### ISSUE THREAD

**Issue by alvinsj** [fnando/i18n-js#389]
I was looking at the change of 3.0.0.rc12, and noticed there is a custom `preprocessor` added in `engine.rb`(https://github.com/fnando/i18n-js/blob/master/lib/i18n/js/engine.rb#L30). I trace the code a little bit, knowing the usage of it for loading translation before `i18n/filtered.js.erb`. However, I also found https://github.com/fnando/i18n-js/blob/master/lib/i18n/js.rb#L153 is already covering the loading of translations.

Before jumping into a conclusion, I would like to ask if I missed anything about the usage of the preprocessor.

**Comment by PikachuEXE** [comment]
Can you give the link to the `preprocessor` you are talking about?
I am trying to avoid any misunderstanding here.

**Comment by alvinsj** [comment]
here https://github.com/fnando/i18n-js/blob/master/lib/i18n/js/engine.rb#L30

**Comment by PikachuEXE** [comment]
This preprocessor is to let Sprockets know that the file with logical path `i18n/filtered`
"depends on" **all** files specified by paths in `::I18n.load_path`.
So that if the content of any file in `::I18n.load_path` is changed, Sprockets will regenerate the JS file.

**Comment by alvinsj** [comment]
Just done some manual test, but didn't get it work: 

1. In `config/environments/development.rb`, unset debug mode with `config.assets.debug = false`
2. check loaded translations javascript file in the html page
3. verify `translation` exists in javascript console by typing `I18n.translations.en.some_key` 
4. change the `locale` file content.
5. refresh and repeat step 2-3.   
**Result: Translation change is NOT reflected**

Then I add `I18n::JS::Middleware`, repeat steps 1-5 again: 
**Result: It works, translation change is reflected**

Then I add `I18n::JS::Middleware` and remove the whole `register_preprocessor` (pointing to my local i18n-js gem), repeat steps 1-5 again 
**Result: It works, translation change is reflected**

**Comment by PikachuEXE** [comment]
Not sure why you have set `debug` to `false`. It's normally `true` in development
Assuming it's true:
- Pick a translation entry
- Load a page
- Ensure it's present in JS console
- Change the content of that entry
- Reload the page
- Ensure it's present & changed in JS console

**Comment by alvinsj** [comment]
Isn't it cache when debug is true? depend_on is for cache right?

**Comment by PikachuEXE** [comment]
What are you trying to do here?
Figure out the purpose of `preprocessor`?
Test the `preprocessor`?

**Comment by alvinsj** [comment]
To investigate if it is redundant code, because if it's running for every JavaScript asset or running this regardless of using the non-middleware method, then it is redundant, getting rid of it means less code to maintain/ less error prone right? My 2 cents

**Comment by alvinsj** [comment]
Yea, the code should work(in my manual tests too), but I just couldn't get it working. 

I mean because of you have middleware method to take care the on-demand code generation, and non-middleware method(`rake I18n:js:export`), looks like both actually don't need the preprocessor. however, I wanted to find out your reason what issue you might have encounter before, that's why you added both.

New finding, it's also seems like it is cached (the file reflects changes but not in javascript environment), even when debug is on. Using the middleware instead seems to able to make it reflect in javascript env. or maybe it's just my application, Im not sure.

**Comment by alvinsj** [comment]
I have the debug OFF, preprocessor REMOVED and CHANGING the locale files. I noticed file fingerprint(and content) still changes accordingly, (even though the locales in javascript env still doesn't change ==). It looks like the content of the file(`i18n/filtered`) is always checked to validate/invalidate the fingerprint too.

on i18n-js 3.0.0.rc12, sprockets 3

**Comment by PikachuEXE** [comment]
Be reminded that there is a third way: using assets pipeline
Using that way will use neither the middleware nor `rake I18n:js:export`

**Comment by alvinsj** [comment]
ok i just tested with the asset pipeline method, following the instruction in the README.md. removing all my code in `application.js.coffee`, adding `#= require 'i18n/translations'` and check on the browsers with the same steps again. surprisingly, this doesn't give me any cache problem, both files and javascript env reflect my locale file change. 

so next I test whether the `preprocessor` code is needed. 
I removed the `preprocessor` code, I tested both condition of setting `config.assets.debug` to `false` and `true`, both reflects the changes, and it still works as well. 

ok you might want to try this at your side, or let me know if there is edge case.
